### ISSUE TITLE
bamsormadup hangs on specific bam

### ISSUE THREAD

**Issue by mpinese** [gt1/biobambam2#23]
Hi German,

I've been using biobambam2 for a while now and it's been a very fast, very solid tool -- thanks!

I've run into a weird problem with bamsormadup on one particular bam, and am wondering how I should debug it.  For just one 165 GB bam generated by bwa mem, bamsormadup seems to hang during the duplicate marking step.  bamsormadup has worked fine for all of the other bams in the same project (which go up to 158 GB in size), and re-mapping the problem bam (just in case odd random file corruption was to blame) hasn't helped.

The tail of the bamsormadup log is:

```
[V] 1906565714  01:10:34:63528599       MemUsage(size=8280.09,rss=7356.48,peak=8724.07) AutoArrayMemUsage(memusage=7351.46,peakmemusage=7466.17,maxmem=1.75922e+13)
[V] 1908071306  01:10:39:29409300       MemUsage(size=8280.09,rss=7356.48,peak=8724.07) AutoArrayMemUsage(memusage=7351.46,peakmemusage=7466.17,maxmem=1.75922e+13)
[V] 1908823598  01:10:40:24747099       MemUsage(size=8280.09,rss=7356.48,peak=8724.07) AutoArrayMemUsage(memusage=7351.55,peakmemusage=7466.17,maxmem=1.75922e+13)
[V] 1909575594  01:10:41:84439600       MemUsage(size=8280.09,rss=7356.48,peak=8724.07) AutoArrayMemUsage(memusage=7351.46,peakmemusage=7466.17,maxmem=1.75922e+13)
[V] 1910203706  01:10:46:07156000       MemUsage(size=8280.09,rss=7356.48,peak=8724.07) AutoArrayMemUsage(memusage=7351.46,peakmemusage=7466.17,maxmem=1.75922e+13)     final
[V] flushing read ends lists...done.
[V] merging read ends lists/computing duplicates...=>> PBS: job killed: walltime 10868 exceeded limit 10800
```

I've tried running this job for up to 12 hours with no luck.  The process was given 16 cores, 24 GB RAM, and 400 GB scratch space; it used at most 7.2 GB RAM and 240 GB scratch.

Do you have any suggestions on how I can debug this?  Are there any special flags to give to bamsormadup that might produce a more verbose log?  The command used was:

```
bamsormadup threads=8 M=SK13-2B.sorted.markdup.txt tmpfile=${PBS_JOBFS}/ indexfilename=SK13-2B.sorted.markdup.bam.bai < SK13-2B.bam > SK13-2B.sorted.markdup.bam
```

Thanks again for your work,

Mark

**Comment by gt1** [comment]
Hi Mark,

in the ideal case you could compile libmaus2 and biobambam2 using --enable-debug during configure and run the command in gdb to see where it gets stuck.

If this is not possible then it would be helpful for me to know a bit more about where it hangs. I have added an option in bamsormadup to print some more information (set blocksortverbose=1 ).

German

**Comment by mpinese** [comment]
Hi German,

Thanks for adding that flag.  I've re-run the job, and the log now looks like:

```
[V] 1508637     02:68878500     MemUsage(size=4099.89,rss=1280.47,peak=4340.04) AutoArrayMemUsage(memusage=3001.61,peakmemusage=3245.45,maxmem=1.75922e+13)
[V] 2263392     03:48657500     MemUsage(size=4099.92,rss=1352.54,peak=4340.04) AutoArrayMemUsage(memusage=3001.66,peakmemusage=3245.45,maxmem=1.75922e+13)
[V] 3773446     05:48968400     MemUsage(size=4426.97,rss=2043.2,peak=4592.26)  AutoArrayMemUsage(memusage=3332.85,peakmemusage=3487.39,maxmem=1.75922e+13)
[V] 4528644     06:85843199     MemUsage(size=4713.95,rss=2343.82,peak=4954.57) 

... clipped ...

[V] 1908823598  01:06:58:00425099       MemUsage(size=8398.96,rss=7392.91,peak=8497.2)  AutoArrayMemUsage(memusage=7351.45,peakmemusage=7414.23,maxmem=1.75922e+13)
[V] 1909575594  01:06:59:71790499       MemUsage(size=8398.96,rss=7392.91,peak=8497.2)  AutoArrayMemUsage(memusage=7351.55,peakmemusage=7414.23,maxmem=1.75922e+13)
[V] 1910203706  01:07:01:86669299       MemUsage(size=8398.96,rss=7392.91,peak=8497.2)  AutoArrayMemUsage(memusage=7351.45,peakmemusage=7414.23,maxmem=1.75922e+13)     final
[V] flushing read ends lists...done.
[V] merging read ends lists/computing duplicates...
        (call enqueMergeFragReadEndsLists)
                (call getFragMergeInfo)
                (return getFragMergeInfo)
                        ([0]=ReadEndsBlockDecoderBaseCollectionInfoBase(datafilename=/jobfs/local/7220819.r-man2/_readends_frag_000004,indexfilename=/jobfs/local/7220819.r-man2/_readends_frag_000004.index,blockelcnt={289268;289268;289268;289267;289267;289267;289267;289266;289267;289266;289266;289267;

... clipped ...

36361315;36406015;36450715;36495415;36540115}))
                (call getShortMergeIntervals)
                (return getShortMergeIntervals)
                        (
                                (0,37354)
                                (0,37440)
                                (0,37105)

 ...clipped...

                                (232045,266305)
                                (232280,266305)
                                (28678,32841)
                        )
                (enque PairReadEndsMergeWorkPackage)
                (done queuing PairReadEndsMergeWorkPackage)
        (return enqueMergePairReadEndsLists)
        (waiting)
=>> PBS: job killed: walltime 10867 exceeded limit 10800
```

Does that help at all?

I've attached the full output log and PBS report here:
[snakejob.bam_sormadup.0.sh.o7220819.txt](https://github.com/gt1/biobambam2/files/338271/snakejob.bam_sormadup.0.sh.o7220819.txt)
[snakejob.bam_sormadup.0.sh.e7220819.zip](https://github.com/gt1/biobambam2/files/338272/snakejob.bam_sormadup.0.sh.e7220819.zip)

**Comment by gt1** [comment]
Hi,

thank you for the logs. It gives me an idea of where it does not hang. Could you please retry with the latest version (just updated) still using the blocksortverbose option? It prints some more useful information.

Thanks
German

**Comment by mpinese** [comment]
Done.  This time I gave the job 6 hours, just to be very sure that it wasn't a time problem.  And this time, it produced an output file!  Unfortunately the job was killed before the bam could be written completely, but I'm going to try again with a very long walltime limit and see.  Why it didn't finish when I gave it 12 hours the first time, or why this particular file takes so much longer than all the others did, I have no idea.

Hopefully this will resolve it.  I've attached the logs just for completeness; thanks for your help so far.

[bam_sormadup_log.zip](https://github.com/gt1/biobambam2/files/340477/bam_sormadup_log.zip)

**Comment by mpinese** [comment]
Hi German,

The longer time limit resolved it; here is the tail of the log:

```
[V] 1910203706
[D]     md5     c293e5dc53432896c97e8ab15f04933c
[V] checksum ok
[V] blocks merged in time 03:25:26:77238899
[V] run time 07:23:22:51451599 (26602.5 s)      MemUsage(size=15541.3,rss=14272.5,peak=21896.2)
```

Sorry about leading you down the garden path on this one.  I have no clue why the jobs were so much slower when I first tried them; perhaps the node I was using at the time was under heavy load.  Thanks for your help, and also again for making this great toolkit!
